nexus/db-queries/src/db/datastore/sled.rs

```diff
@@ -564,14 +736,370 @@ impl DataStore {
+            } else {
+                // If local storage allocation is required, match the requests
+                // with all the zpools of this sled that have available space.
+                // This routine finds all possible configurations that would
+                // satisfy the requests for local storage and tries them all.
+
+                // First, each request for local storage can possibly be
+                // satisfied by a number of zpools on the sled. Find this list
+                // of candidate zpools for each required local storage
+                // allocation.
+
+                #[derive(Clone, Debug, Hash, PartialEq, Eq)]
+                struct CandidateDataset {
+                    rendezvous_local_storage_dataset_id: DatasetUuid,
+                    pool_id: ZpoolUuid,
+                    sled_id: SledUuid,
+                }
+
+                #[derive(Clone, Debug)]
+                struct PossibleAllocationsForRequest<'a> {
+                    request: &'a LocalStorageDisk,
+                    candidate_datasets: HashSet<CandidateDataset>,
+                }
+
+                let zpools_for_sled = self
+                    .zpool_get_for_sled_reservation(&opctx, sled_target)
+                    .await?;
+
+                // If there's an existing local storage allocation on a zpool,
+                // remove that from the list of candidates. Local storage for
+                // the same Instance should not share any zpools.
+                let local_storage_zpools_used: HashSet<ZpoolUuid> =
+                    local_storage_disks
+                        .iter()
+                        .filter_map(|disk| {
+                            disk.local_storage_dataset_allocation.as_ref().map(
+                                |allocation| {
+                                    ZpoolUuid::from_untyped_uuid(
+                                        allocation
+                                            .pool_id()
+                                            .into_untyped_uuid(),
+                                    )
+                                },
+                            )
+                        })
+                        .collect();
+
+                let zpools_for_sled: Vec<_> = zpools_for_sled
+                    .into_iter()
+                    .filter(|zpool_get_result| {
+                        !local_storage_zpools_used
+                            .contains(&zpool_get_result.pool.id())
+                    })
+                    .collect();
+
+                if local_storage_allocation_required.len()
+                    > zpools_for_sled.len()
+                {
+                    // Not enough zpools to satisfy the number of allocations
+                    // required. Find another sled!
+                    info!(
+                        &log,
+                        "sled {sled_target:?} does not have enough zpools to \
```

Comment:
wildly unimportant nit: the `Debug` formatter for `SledUuid` will print `$(uuidgen -r) (sled)`, while the `Display` formatter will just print `"$(uuidgen -r)`. might read better if we changed it to
```suggestion
                        "sled {sled_target} does not have enough zpools to \
```

..or, perhaps, record the sled ID as a structured field?